hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/message/MessageState.java

```diff
@@ -1,33 +1,53 @@
 package pl.allegro.tech.hermes.frontend.publishing.message;
 
-import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.SENDING_TO_KAFKA;
-import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.SENT_TO_KAFKA;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.READING;
 
 public class MessageState {
 
     public enum State {
-        WAITING_FOR_FIRST_PORTION_OF_DATA,
-        PARSING,
-        PARSED,
+        READING,
+        READING_TIMEOUT,
+        FULLY_READ,
         SENDING_TO_KAFKA_PRODUCER_QUEUE,
+        ERROR_IN_SENDING_TO_KAFKA_PRODUCER_QUEUE,
         SENDING_TO_KAFKA,
-        SENT_TO_KAFKA
+        SENT_TO_KAFKA,
+        DELAYED_SENDING,
+        DELAYED_PROCESSING
     }
 
-    private volatile State state = State.WAITING_FOR_FIRST_PORTION_OF_DATA;
+    private final String messageId;
 
     public MessageState() {
+        messageId = MessageIdGenerator.generate();
+    }
+
+    private volatile boolean delayed = false;
+    private AtomicReference<State> state = new AtomicReference<>(READING);
+
+    public boolean compareAndSet(State expect, State update) {
+        return state.compareAndSet(expect, update);
+    }
+
+    public void set(State state) {
+        this.state.set(state);
+    }
+
+    public State get() {
+        return state.get();
     }
 
-    public synchronized void setState(State state) {
-        this.state = state;
+    public boolean isDelayed() {
+        return delayed;
```

Comment:
Ok I get this but what I don't like is that transitions and validation between states are done outside of state machine itself (I mean MessageState). In my opinion other classes should try to advance state to some desired value but state machine should validate itself whether it is possible to do that and return feedback if it is possible. Right now validation between states is scattered around handlers which makes the whole thing hard to understand and hard to modify.

Example:
```
try {
   state.advance(someState);
   doSomethingEpic();
} catch (IllegalThingGoingOn ex) {
   doSomethingDifferent();
}
```